File: GithubUserJava/app/src/main/java/com/example/myapplication/ui/FavoriteActivity.java
package com.example.myapplication.ui;

import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.myapplication.adapter.FavoriteAdapter;
import com.example.myapplication.databinding.ActivityFavoriteBinding;
import com.example.myapplication.factory.FavoriteViewModelFactory;
import com.example.myapplication.viewmodel.FavoriteViewModel;

public class FavoriteActivity extends AppCompatActivity {

    private ActivityFavoriteBinding binding;
    private FavoriteViewModel favoriteViewModel;
    private final FavoriteAdapter adapter = new FavoriteAdapter();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityFavoriteBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        favoriteViewModel = obtainViewModel(this);
        setUpList();
        setUserFavorite();
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    private FavoriteViewModel obtainViewModel(AppCompatActivity activity) {
        FavoriteViewModelFactory factory = FavoriteViewModelFactory.getInstance(activity.getApplication());
        return new ViewModelProvider(activity, factory).get(FavoriteViewModel.class);
    }

    private void setUpList() {
        binding.rvFavorite.setLayoutManager(new LinearLayoutManager(this));
        DividerItemDecoration itemDecoration = new DividerItemDecoration(this, LinearLayoutManager.VERTICAL);
        binding.rvFavorite.addItemDecoration(itemDecoration);
        binding.rvFavorite.setAdapter(adapter);

        adapter.setOnItemClickCallback(favEntity -> {
            Intent intent = new Intent(FavoriteActivity.this, DetailActivity.class);
            intent.putExtra(DetailActivity.KEY_USERNAME, favEntity.getUsername());
            startActivity(intent);
        });
    }

    private void setUserFavorite() {
        favoriteViewModel = obtainViewModel(this);
        favoriteViewModel.getAllFavorites().observe(this, favList -> {
            if (favList != null) {
                adapter.setListFavorite(favList);
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding = null;
    }
}

File: GithubUserKotlin/app/src/main/java/com/example/myapplication/ui/FavoriteActivity.kt
package com.example.myapplication.ui

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.myapplication.adapter.FavoriteAdapter
import com.example.myapplication.data.local.entity.FavoriteEntity
import com.example.myapplication.databinding.ActivityFavoriteBinding
import com.example.myapplication.factory.FavoriteViewModelFactory
import com.example.myapplication.viewmodel.FavoriteViewModel


class FavoriteActivity : AppCompatActivity() {
    private var _binding: ActivityFavoriteBinding? = null
    private val binding get() = _binding
    private lateinit var favoriteViewModel: FavoriteViewModel
    private val adapter: FavoriteAdapter by lazy {
        FavoriteAdapter()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityFavoriteBinding.inflate(layoutInflater)
        setContentView(binding?.root)
        favoriteViewModel = obtainViewModel(this@FavoriteActivity)
        setUpList()
        setUserFavorite()
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }
    private fun obtainViewModel(activity: AppCompatActivity): FavoriteViewModel {
        val factory = FavoriteViewModelFactory.getInstance(activity.application)
        return ViewModelProvider(activity, factory)[FavoriteViewModel::class.java]
    }

    private fun setUpList() {
        with(binding) {
            val layoutManager = LinearLayoutManager(this@FavoriteActivity)
            this?.rvFavorite?.layoutManager = layoutManager
            val itemDecoration =
                DividerItemDecoration(this@FavoriteActivity, layoutManager.orientation)
            this?.rvFavorite?.addItemDecoration(itemDecoration)
            this?.rvFavorite?.adapter = adapter
            adapter.setOnItemClickCallback(object : FavoriteAdapter.OnItemClickCallback {
                override fun onItemClicked(favEntity: FavoriteEntity) {
                    val intent = Intent(this@FavoriteActivity, DetailActivity::class.java)
                    intent.putExtra(DetailActivity.KEY_USERNAME, favEntity.username)
                    startActivity(intent)
                }
            })
        }
    }
    private fun setUserFavorite() {
        favoriteViewModel = obtainViewModel(this@FavoriteActivity)
        favoriteViewModel.getAllFavorites().observe(this@FavoriteActivity) { favList ->
            if (favList != null) {
                adapter.setListFavorite(favList)
            }
        }
    }
    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }
}